matchzoo/generators/list_generator.py

```diff
@@ -0,0 +1,131 @@
+        ...             ['qid0', 'did1', 1],
+        ...             ['qid0', 'did2', 2]
+        ... ]
+        >>> left = [['qid0', [1, 2]]]
+        >>> right = [['did0', [2, 3]],
+        ...          ['did1', [3, 4]],
+        ...          ['did2', [4, 5]],
+        ... ]
+        >>> relation = pd.DataFrame(relation,
+        ...                         columns=['id_left', 'id_right', 'label'])
+        >>> left = pd.DataFrame(left, columns=['id_left', 'text_left'])
+        >>> left.set_index('id_left', inplace=True)
+        >>> right = pd.DataFrame(right, columns=['id_right', 'text_right'])
+        >>> right.set_index('id_right', inplace=True)
+        >>> input = datapack.DataPack(relation=relation,
+        ...                           left=left,
+        ...                           right=right
+        ... )
+        >>> generator = ListGenerator(input)
+        >>> len(generator)
+        1
+        >>> x, y = generator[0]
+        >>> x['text_left'].tolist()
+        [[1, 2], [1, 2], [1, 2]]
+        >>> x['text_right'].tolist()
+        [[2, 3], [3, 4], [4, 5]]
+        >>> x['ids'].tolist()
+        [['qid0', 'did0'], ['qid0', 'did1'], ['qid0', 'did2']]
+        >>> y.tolist()
+        [0.0, 1.0, 2.0]
+
+    """
+
+    def __init__(
+        self,
+        inputs: datapack.DataPack,
+        batch_size: int=1,
+        stage: str='train',
+        shuffle: bool=True
+    ):
+        """Construct the list generator.
+
+        :param inputs: the output generated by :class:`DataPack`.
+        :param batch_size: number of instances in a batch.
+        :param stage: the current phase, the value can be 'train' or 'test'.
+        :param shuffle: whether to shuffle the instances while generating a
+            batch.
+        """
+        self._left = inputs.left
+        self._right = inputs.right
+        self._relation = inputs.relation
+        self._task = tasks.Ranking()
+        self._lists = self.transform_relation(self._relation)
+        super().__init__(batch_size, len(self._lists), stage, shuffle)
+
+    def transform_relation(self, relations: pd.DataFrame) -> list:
+        """Obtain the transformed data from :class:`DataPack`.
+
+        Note here, label is required to make lists.
+
+        :param relations: An instance of DataFrame to be transformed.
+        :return: the output of all the lists' indices.
+        """
+        # Note here the main id is set to be the id_left
+        lists = []
+        for idx, group in relations.groupby('id_left'):
+            lists.append(group.index.tolist())
+        return lists
+
+    def _get_batch_of_transformed_samples(
+        self,
+        index_array: np.array
+    ) -> typing.Tuple[dict, typing.Any]:
+        """Get a batch of samples based on their ids.
+
+        :param index_array: a list of instance ids.
+        :return: A batch of transformed samples.
+        """
+        trans_index = self._lists[index_array[0]]
```

Comment:
not clear what you're doing here

Reply:
Right now in this version~ The batch_size is set to be 1. Thus, the `index_array` is a list which contains only one element.